File: bindings/python/aer.py
"""Minimal Python AER text emitter for interoperability experiments."""
import json
from typing import Any


def scalar(value: Any) -> str:
    if value is None:
        return "-"
    if isinstance(value, bool):
        return str(value).lower()
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, str):
        return json.dumps(value) if any(c in value for c in ',:\n"{}[]') else value
    return json.dumps(value, separators=(",", ":"))


def encode(value: dict[str, Any], indent: str = "") -> str:
    lines: list[str] = []
    for key, item in value.items():
        if isinstance(item, list) and item and all(isinstance(x, dict) for x in item):
            columns: list[str] = []
            for row in item:
                for column in row:
                    if column not in columns:
                        columns.append(column)
            lines.append(f"{indent}{key}[{len(item)}]{{{','.join(columns)}}}:")
            for row in item:
                lines.append(indent + "  " + ",".join(scalar(row.get(c)) for c in columns))
        elif isinstance(item, dict):
            lines.append(f"{indent}{key}:")
            lines.extend(encode(item, indent + "  ").splitlines())
        elif isinstance(item, list):
            lines.append(f"{indent}{key}[{len(item)}]:" + ",".join(scalar(x) for x in item))
        else:
            lines.append(f"{indent}{key}:{scalar(item)}")
    return "\n".join(lines)

File: bindings/typescript/aer.ts
export function scalar(value: unknown): string {
  if (value === null || value === undefined) return "-";
  if (typeof value === "boolean") return String(value);
  if (typeof value === "number") return String(value);
  if (typeof value === "string") return /[,;:\n"{}\[\]]/.test(value) ? JSON.stringify(value) : value;
  return JSON.stringify(value);
}

export function encode(value: Record<string, unknown>, indent = ""): string {
  const lines: string[] = [];
  for (const [key, item] of Object.entries(value)) {
    if (Array.isArray(item) && item.length && item.every(x => x && typeof x === "object" && !Array.isArray(x))) {
      const columns: string[] = [];
      for (const row of item as Record<string, unknown>[]) for (const column of Object.keys(row)) if (!columns.includes(column)) columns.push(column);
      lines.push(`${indent}${key}[${item.length}]{${columns.join(",")}}:`);
      for (const row of item as Record<string, unknown>[]) lines.push(`${indent}  ${columns.map(c => scalar(row[c])).join(",")}`);
    } else if (item && typeof item === "object" && !Array.isArray(item)) {
      lines.push(`${indent}${key}:`);
      lines.push(...encode(item as Record<string, unknown>, indent + "  ").split("\n"));
    } else if (Array.isArray(item)) {
      lines.push(`${indent}${key}[${item.length}]:${item.map(scalar).join(",")}`);
    } else {
      lines.push(`${indent}${key}:${scalar(item)}`);
    }
  }
  return lines.join("\n");
}
